twilio/base/client_base.py

```diff
@@ -85,8 +88,21 @@ def request(
 
         :returns: Response from the Twilio API
         """
-        auth = self.get_auth(auth)
+
         headers = self.get_headers(method, headers)
+
+        ##If credential provider is provided by user, get the associated auth strategy
+        ##Using the auth strategy, fetch the auth string and set it to authorization header
+        auth_strategy = None ##Initialization
+        if self.credential_provider:
+            auth_strategy = self.credential_provider.to_auth_strategy()
+            if auth_strategy.auth_type == AuthType.TOKEN:
```

Comment:
This works here, but we should avoid checking if conditions for auth_type. Instead. we can create an abstract function at auth_strategy class that defines how each auth strategy sets header for itself. Let's say that function is `setAuth`. This setAuth() for Token authtype will call fetch_token() and set authorization header.

So, from here, we just need to call auth_strategy.setAuth() and all the things will be abstracted at the implementation level.

Reply:
Moved this code to auth strategy